pages/api/applications/[applicationId]/stages/index.ts

```diff
@@ -1,5 +1,146 @@
+import { isValidDate } from '../../../../../utils/date/validations';
+import { withAuthUser } from '../../../../../utils/auth/jwtHelpers';
+import { Nullable } from '../../../../../types/utils';
+import { ApplicationDataWithStagesOnly } from '../../../../../types/application';
+import { isValidHex } from '../../../../../utils/strings/validations';
+import { canBecomeInteger } from '../../../../../utils/numbers/validations';
 
-export default function handler(req: NextApiRequest, res: NextApiResponse) {
-  res.status(200).end(`Request sent to ${req.url}\n`);
+const prisma = new PrismaClient();
+
+enum MessageType {
+  INVALID_TYPE,
+  INVALID_DATE,
+  INVALID_APPLICATION_ID,
+  APPLICATION_NOT_FOUND,
+  CREATED,
+  INVALID_EMOJI_UNICODE_HEX,
+  INVALID_REMARK,
+}
+
+const messages = Object.freeze({
+  [MessageType.INVALID_TYPE]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage type is invalid.',
+  },
+  [MessageType.INVALID_DATE]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage date is invalid.',
+  },
+  [MessageType.INVALID_APPLICATION_ID]: {
+    type: StatusMessageType.ERROR,
+    message: 'ApplicationId is invalid.',
+  },
+  [MessageType.APPLICATION_NOT_FOUND]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application cannot be found.',
+  },
+  [MessageType.CREATED]: { type: StatusMessageType.SUCCESS, message: 'Application stage has been created.' },
+  [MessageType.INVALID_REMARK]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage remark is invalid.',
+  },
+  [MessageType.INVALID_EMOJI_UNICODE_HEX]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage emoji unicode hex is invalid.',
+  },
+});
+
+async function handler(
+  uid: string,
+  req: NextApiRequest,
+  res: NextApiResponse<ApiResponse<ApplicationStageData | ApplicationDataWithStagesOnly>>,
+) {
+  switch (req.method) {
+    case HttpMethod.POST:
+      return handlePost(uid, req, res);
+    default:
+      return rejectHttpMethod(res, req.method);
+  }
+}
+
+async function handlePost(uid: string, req: NextApiRequest, res: NextApiResponse<ApiResponse<ApplicationStageData>>) {
+  const paramValidationError = validatePostRequest(req);
+  if (paramValidationError !== null) {
+    return res.status(HttpStatus.BAD_REQUEST).json(createJsonResponse({}, messages[paramValidationError]));
+  }
+
+  const application = await getApplication(uid, Number(req.query.applicationId));
+  if (application === null) {
+    return res.status(HttpStatus.NOT_FOUND).json(createJsonResponse({}, messages[MessageType.APPLICATION_NOT_FOUND]));
+  }
+
+  const { type, date, emojiUnicodeHex, remark } = req.body;
+  // Empty or undefined emojis and remarks will be set to null.
+  const parameters: ApplicationStagePostData = {
+    type,
+    date: new Date(date),
+    emojiUnicodeHex: emojiUnicodeHex || null,
+    remark: remark || null,
```

Comment:
```suggestion
    emojiUnicodeHex: emojiUnicodeHex,
    remark: remark,
```

I don't think we need this. I tested at it works without.